community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py

```diff
@@ -472,6 +485,129 @@ def update_topology(lkp: util.Lookup) -> None:
+    if res.returncode != 0:
+        raise subprocess.SubprocessError(f"{err_msg}:\n{res.stderr}")
+
+    return res
+
+def delete_reservations(reservation_map: Dict[str, str]) -> None:
+    err = "error deleting reservation"
+    for reservation_name in list(reservation_map.keys()):
+      res = run_command(SLURM_DELETE_RESERVATION.format(reservation_name), err)
+      del reservation_map[reservation_name]
+
+
+def update_slurm_reservation_maintenance(reservation_name: str, node: str,
+                                         starttime: str, reservation_map: Dict[str, str]) -> None:
+    ## If reservation exists for maintenance, update reservation if time window
+    ## for scheduled maintenance changes.
+    if reservation_map is not None and reservation_name in reservation_map:
+      details = reservation_map[reservation_name]
+
+      nodes = None
+      start_time = None
+
+      values = details.split(",")
+      nodes = values[0].strip()
+      maintenance_start_time = values[1].strip()
+
+      if nodes and start_time:
+        if maintenance_start_time == starttime:
+          return
+
+        err = "error deleting reservation"
+        res = run_command(SLURM_DELETE_RESERVATION.format(reservation_name), err)
+
+        err = "error updating reservation"
+        res = run_command(SLURM_CREATE_RESERVATION.format(starttime, nodes, reservation_name), err)
+
+      del reservation_map[reservation_name]  # remove the reservation from the map.
+
+    else: # Reservation doesn't exist, make reservation during maintenance window.
+      err = "error creating reservation."
+      res = run_command(SLURM_CREATE_RESERVATION.format(starttime, node, reservation_name), err)
+
+
+def get_slurm_reservation_maintenance() -> Dict[str, str]:
+    err = "error getting reservation details"
+    res = run_command(SLURM_SHOW_RESERVATION, err)
+    all_reservations = [x.split() for x in res.stdout.split("\n\n")[:-1]]
+    reservation_map = {}
+
+    for reservation in all_reservations:
+        reservation_name = None
+        nodes = None
+        start_time = None
+        for item in reservation:
+          key, value = item.split('=', 1)  # Split at the first '='
+
+          if key == 'ReservationName':
+            reservation_name = value
+          elif key == 'Nodes':
+            nodes = value
+          elif key == 'StartTime':
+            start_time = value
+
+        if reservation_name is None or nodes is None or start_time is None:
+          continue
+
+        # Check if the reservation is for scheduled maintenance.
+        maintenance_reservation = nodes + "_maintenance"
+        if reservation_name != maintenance_reservation:
+          continue
+
+        reservation_map[reservation_name] = nodes + ', ' + start_time
+
+    return reservation_map
+
+
+def update_upcoming_maintenance(lkp: util.Lookup) -> List[str]:
+    err = "error getting upcoming maintenance list"
+    res = run_command(UPC_MAINT_CMD.format(lkp.project), err)
```

Comment:
We probably don't want to run this command. Let's use already fetched info, see `lkp.instances()`

Reply:
Updated. Currently we use `beta` version for compute and upcoming maintenance is not supported in beta yet. Updated util script to add upcoming maintenance notification information if api version is alpha.